File: razer-sdk/src/main/java/com/bonepl/razersdk/animation/AnimatedFrame.java
package com.bonepl.razersdk.animation;

import java.util.LinkedList;
import java.util.NoSuchElementException;
import java.util.Queue;
import java.util.stream.IntStream;

/**
 * Object containing a sequence of {@link SimpleFrame} that build the animation.
 * Each following {@link Frame} is polled by calling {@link #getFrame()}.
 * Speed of the animation will depend on the external scheduler calling {@link #getFrame()}.
 */
public class AnimatedFrame implements IFrame {
    public final Queue<Frame> frames = new LinkedList<>();

    /**
     * Add single animation frame delivered by calling {@link IFrame#getFrame()}
     *
     * @param frame frame added to animation sequence
     */
    public final void addAnimationFrame(IFrame frame) {
        addAnimationFrame(1, frame);
    }

    /**
     * Add single animation frame extended by frames count.
     * Having external animation scheduler, this is used to set the length of current's
     * frame display. For example if you get animation frame every 50ms and you want
     * the frame to last 200ms set frameCount to 4.
     *
     * @param frameCount how long should provided frame last (be returned by {@link #getFrame()}
     * @param frame      frame to be added to animation sequence
     */
    public final void addAnimationFrame(int frameCount, IFrame frame) {
        if (frameCount > 0) {
            final Frame animationFrame = frame.getFrame();
            IntStream.range(0, frameCount).forEach(i -> frames.offer(animationFrame));
        }
    }

    /**
     * Check if frame is available
     *
     * @return true if there are still frames in the animation sequence,
     * false otherwise
     */
    @Override
    public boolean hasFrame() {
        return !frames.isEmpty();
    }

    /**
     * Return the frame
     *
     * @return next {@link Frame} of the animation
     * @throws NoSuchElementException when there is no frames available
     */
    @Override
    public Frame getFrame() {
        if (hasFrame()) {
            return frames.remove();
        } else {
            throw new NoSuchElementException("AnimatedFrame does not have any frames to return");
        }
    }

    /**
     * Clear the frame sequence, and stop the animation
     */
    public final void clearFrames() {
        frames.clear();
    }
}

File: src/razer-sdk/animation/AnimatedFrame.ts
import {Frame} from './Frame.js';
import {IFrame} from './IFrame.js';

/**
 * Object containing a sequence of Frames that build the animation.
 * Each following Frame is polled by calling pollFrame().
 * Speed of the animation depends on the external scheduler calling pollFrame().
 */
export class AnimatedFrame implements IFrame {
    private readonly frames: Frame[] = [];

    /** Add single animation frame */
    addAnimationFrame(frame: IFrame): void;
    /** Add single animation frame extended by frameCount */
    addAnimationFrame(frameCount: number, frame: IFrame): void;
    addAnimationFrame(frameCountOrFrame: number | IFrame, frame?: IFrame): void {
        let frameCount: number;
        let actualFrame: IFrame;

        if (typeof frameCountOrFrame === 'number') {
            frameCount = frameCountOrFrame;
            actualFrame = frame!;
        } else {
            frameCount = 1;
            actualFrame = frameCountOrFrame;
        }

        if (frameCount > 0) {
            const animationFrame = actualFrame.pollFrame();
            for (let i = 0; i < frameCount; i++) {
                this.frames.push(animationFrame);
            }
        }
    }

    hasFrame(): boolean {
        return this.frames.length > 0;
    }

    pollFrame(): Frame {
        if (this.hasFrame()) {
            return this.frames.shift()!;
        }
        throw new Error('AnimatedFrame does not have any frames to return');
    }

    /** Clear the frame sequence, and stop the animation */
    clearFrames(): void {
        this.frames.length = 0;
    }
}
